### PR TITLE
Add documentation for experimental features.

### PR DESCRIPTION
**Pull request by clokep**
Per our discussions of when to use experimental config flags.

### CODE REVIEW

File: docs/development/experimental_features.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Implementing experimental features in Synapse
+
+It can be desirable to implement "experimental" features which are disabled by
+default and must be explicitly enabled via the Synapse configuration. This is
+applicable for features which:
+
+* Are unstable in the Matrix spec (e.g. those defined by an MSC).
```
Comment (anoadragon453) [inline docs/development/experimental_features.md:7]:
```suggestion
* Are unstable in the Matrix spec (e.g. those defined by an MSC that has not yet been merged).
```

File: docs/development/experimental_features.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+* Developers are not confident are ready for use by general Synapse administrators/users
```
Comment (anoadragon453) [inline docs/development/experimental_features.md:8]:
```suggestion
* Developers are not confident in their use by general Synapse administrators/users
```

Possibly?

File: docs/development/experimental_features.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  (e.g. a feature is incomplete, buggy, performs poorly, or needs further testing).
+
+Note that this does *not* include features which are not expected to be desirable
```
Comment (anoadragon453) [inline docs/development/experimental_features.md:11]:
```suggestion
Note that this only really applies to features which are expected to be desirable
```

An attempt at eliminating double negatives.

File: docs/development/experimental_features.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+to a broad audience. The [module infrastructure](../modules/index.md) should be
+investigated for non-standard features.
```
Comment (anoadragon453) [inline docs/development/experimental_features.md:13]: Is the suggestion here that some module features should be hidden behind experimental config flags, or?

File: docs/development/experimental_features.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+* Ease testing a feature (or removal of feature) due to enabling/disabling without
+  code changes. It also becomes possible to ask for wider testing, if desired.
+
+Experimental configuration flags should be disabled by default (requiring Synapse
+administrators to explicitly opt-in), although there are situations where it makes
+sense (from a product point-of-view) to enable features by default. This is
+expected and not an issue.
+
+It is not a requirement that experimental features require a configuration flag,
+but if one should be included if unsure.
```
Comment (anoadragon453) [inline docs/development/experimental_features.md:33]:
```suggestion
It is not a requirement for experimental features to be behind a configuration flag,
but one should be used if unsure.
```

**Review by anoadragon453 — APPROVED** [review]
Thank you so much for documenting this! I just had a few wording nits below, but otherwise this looks like it covers all the usual questions (other than documenting them separately).

File: docs/development/experimental_features.md (hunk 4 above)
Comment (clokep) [reply docs/development/experimental_features.md:13]: The suggestion is that some features should be built as modules, not as things behind experimental config flags.

Comment (clokep) [reply docs/development/experimental_features.md:13]: Maybe it is clearer to just remove this? Not sure, was trying to give people other options though.

Comment (anoadragon453) [reply docs/development/experimental_features.md:13]: Aha OK! In that case, I might just slightly reword this to:

```suggestion
to a broad audience. The [module infrastructure](../modules/index.md) should
instead be investigated for non-standard features.
```
